### PR TITLE
IA-4720 Calculate spark.driver.memory

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/IA-4720

The PR assumes that for existing AoU image, the new algorithm will work fine as well

### What

-

### Why

-

### What to test

- 

### Who tested and where

- [ ] This change is covered by automated tests 
  - _NB: Rerun automation tests on this PR by commenting `jenkins retest` or `jenkins multi-test`._
- [ ] I validated this change 
- [ ] Primary reviewer validated this change 
- [ ] I validated this change in the dev environment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala [hunk 1]
```diff
@@ -786,13 +786,21 @@ class DataprocInterpreter[F[_]: Parallel](
       val sparkMemoryConfigRatio = config.dataprocConfig.sparkMemoryConfigRatio.getOrElse(0.8)
       // We still want a minimum to run Jupyter and other system processes.
       val minRuntimeMemoryGb = config.dataprocConfig.minimumRuntimeMemoryInGb.getOrElse(4.0)
+      // Note this algorithm is recommended by Hail team. See more info in https://broadworkbench.atlassian.net/browse/IA-4720
+      val sparkDriverMemoary = machineType match {
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala]: nit, `Memory`

**Review by LizBaldo — APPROVED** [review]
LGTM, thanks for taking care of this :)

I only have a small comment, but in addition of that do you need help with testing on a BEE? We should be able to spin up both a 2.1.15 and 2.2.8 AOU images from the leo swagger page and run a few test commands in the dataproc notebooks.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala [hunk 2]
```diff
@@ -786,13 +786,21 @@ class DataprocInterpreter[F[_]: Parallel](
       val sparkMemoryConfigRatio = config.dataprocConfig.sparkMemoryConfigRatio.getOrElse(0.8)
```
Comment (yonghaoy) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala:786]: Is `sparkMemoryConfigRatio` still used? Shall we change it to 0.9 then reuse this config?

**Review by yonghaoy — APPROVED** [review]

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala (hunk 2 above)
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala:786]: yea, if machine types are not `n1-standard` or `n1-highmem`, it'll use the old algorithem

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala [hunk 3]
```diff
@@ -786,13 +786,21 @@ class DataprocInterpreter[F[_]: Parallel](
...
       // We still want a minimum to run Jupyter and other system processes.
       val minRuntimeMemoryGb = config.dataprocConfig.minimumRuntimeMemoryInGb.getOrElse(4.0)
+      // Note this algorithm is recommended by Hail team. See more info in https://broadworkbench.atlassian.net/browse/IA-4720
+      val sparkDriverMemory = machineType match {
+        case MachineTypeName(n1standard) if n1standard.startsWith("n1-standard") =>
+          Some(MemorySize.fromGb((total.bytes / MemorySize.gbInBytes - 7) * 0.9))
+        case MachineTypeName(n1highmem) if n1highmem.startsWith("n1-highmem") =>
+          Some(MemorySize.fromGb((total.bytes / MemorySize.gbInBytes - 11) * 0.9))
+        case _ => none[MemorySize]
```
Comment (cpate4) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala:795]: I believe these are setting to 90% after carving out something for the OS, maybe? question-mark?

cc: @cindy-broadinstitute

**Comment by Qi77Qi** [comment]
unrelated GCE failures
```
[org.broadinstitute.dsde.workbench.leonardo.runtimes.RuntimeGceSpec.should create a GCE instance in a non-default zone](https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-fiab-test-runner/58766/testReport/junit/org.broadinstitute.dsde.workbench.leonardo.runtimes/RuntimeGceSpec/should_create_a_GCE_instance_in_a_non_default_zone/)
[org.broadinstitute.dsde.workbench.leonardo.runtimes.RuntimePatchSpec.Patch endpoint should perform a stop/start transition for GCE VM with PD](https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-fiab-test-runner/58766/testReport/junit/org.broadinstitute.dsde.workbench.leonardo.runtimes/RuntimePatchSpec/Patch_endpoint_should_perform_a_stop_start_transition_for_GCE_VM_with_PD/)
```

jenkins retest

**Comment by Qi77Qi** [comment]
a lot of failures...seems some GCP issue

jenkins retest

**Pull request merged**